File: chain/tests/initialize.ts
import { expect } from 'chai'
import {
  TOKEN_2022_PROGRAM_ID,
  TOKEN_PROGRAM_ID,
  getMint,
  getPermanentDelegate,
  getTokenMetadata,
} from '@solana/spl-token'
import * as anchor from '@coral-xyz/anchor'
import { readFileSync } from 'node:fs'
import {
  ASSET_COUNT,
  BN,
  GENESIS_M2,
  assetSpec,
  harness,
  initAssetAccounts,
  initParams,
  initializeAccounts,
  TERM_SECONDS,
  EPOCH_SECONDS,
  GRANTS_PER_EPOCH,
} from './enc-harness.js'

/**
 * Bootstrap: the mint, the vault, the rules, and the ten assets.
 *
 * **Written to be re-runnable against a persistent ledger.** `initialize` is
 * once-ever by construction, so this suite creates what is missing and then
 * asserts the invariants on whatever exists. Several tests can only be
 * meaningful on a fresh chain and mark themselves pending otherwise, rather
 * than passing for the wrong reason — `./stack reset` if you want them to run.
 */
describe('initialize + init_asset', () => {
  const h = harness()
  const K = BigInt(h.params.peg.k)
  const EXPECTED_SUPPLY = GENESIS_M2 * K

  const created = new Map<number, string>()

  /**
   * What a stranger got when they tried to bootstrap, attempted *before* the
   * asset existed.
   *
   * The timing is the point. Anchor evaluates `init` during account
   * construction but bare `constraint =` expressions only afterwards, so
   * attempting an index that already exists fails with "account already in use"
   * and never reaches the authority check. The transaction is atomic either
   * way, so the gate holds — but a test aimed at an existing index proves
   * nothing about it.
   */
  let strangerFailure: string | null = null

  before(async function () {
    this.timeout(300_000)

    if (!(await h.exists(h.configPda))) {
      await h.program.methods
        .initialize(initParams(h) as never)
        .accounts(initializeAccounts(h))
        .rpc()
    }

    for (let i = 0; i < ASSET_COUNT; i++) {
      if (await h.exists(h.assetPda(i))) continue
      const spec = assetSpec(i)

      if (strangerFailure === null) {
        const stranger = anchor.web3.Keypair.generate()
        const drop = await h.connection.requestAirdrop(stranger.publicKey, 1e9)
        await h.connection.confirmTransaction(drop)
        strangerFailure = await h.failureOf(() =>
          h.program.methods
            .initAsset(i, spec.name, spec.symbol, spec.uri, spec.price)
            .accounts(initAssetAccounts(h, i, stranger.publicKey))
            .signers([stranger])
            .rpc(),
        )
      }

      created.set(
        i,
        await h.program.methods
          .initAsset(i, spec.name, spec.symbol, spec.uri, spec.price)
          .accounts(initAssetAccounts(h, i))
          .rpc(),
      )
    }
  })

  // ── The rules ─────────────────────────────────────────────────────────────

  it('records the genesis parameters exactly as given', async () => {
    const config = await h.program.account.config.fetch(h.configPda)
    expect(config.k.toString()).to.equal(String(h.params.peg.k))
    expect(config.encDecimals).to.equal(6)
    expect(config.floorBps).to.equal(h.params.vault.floorBps)
    expect(config.faucetAlphaBps).to.equal(h.params.faucet.alphaBps)
    // Not the genesis 30 days: this ledger runs short terms so the auction's
    // settlement paths are reachable at all. See TERM_SECONDS.
    expect(Number(config.termSeconds)).to.equal(TERM_SECONDS)
    // Same substitution, same reasoning: at the shipped one-day epoch not one
    // faucet case is reachable. Asserted rather than assumed, so the swap is
    // visible in the record.
    expect(Number(config.epochSeconds)).to.equal(EPOCH_SECONDS)
    expect(config.grantsPerEpoch).to.equal(GRANTS_PER_EPOCH)
    expect(config.welcomeGrant.toString()).to.equal(String(h.params.faucet.welcomeGrant))
    expect(config.maxChangeBps).to.equal(h.params.sanity.maxChangeBps)
    expect(config.mint.toBase58()).to.equal(h.mintPda.toBase58())
    expect(config.vault.toBase58()).to.equal(h.vaultPda.toBase58())
  })

  it('is callable exactly once', async () => {
    const why = await h.failureOf(() =>
      h.program.methods
        .initialize(initParams(h) as never)
        .accounts(initializeAccounts(h))
        .rpc(),
    )
    expect(why, 'initialize ran a second time — the parameters are not immutable').to.not.equal('')
  })

  /**
   * The parameters are meant to be unchangeable, and the way to be sure is that
   * no instruction to change them exists — not that one exists and is gated.
   * This asserts the shape of the program, not its behaviour.
   */
  it('exposes no instruction that could rewrite the rules', () => {
    // An allowlist rather than a pattern: every new instruction has to be added
    // here deliberately, which is the moment to ask whether it touches Config.
    // (A regex is worse than useless — /set/ matches "initAsset".)
    const allowed = [
      'initialize', 'init_asset', 'initAsset',
      'sync_m2', 'syncM2',
      'place_bid', 'placeBid',
      'withdraw_bid', 'withdrawBid',
      'settle_auction', 'settleAuction',
      'roll_term', 'rollTerm',
      'mint_certificate', 'mintCertificate',
      'claim',
      'close_epoch', 'closeEpoch',
      // Mock builds only; absent from a real one, which the test below proves.
      'set_mock_m2', 'setMockM2',
      'mock_fund', 'mockFund',
    ]
    const unexpected = h.program.idl.instructions
      .map((i) => i.name)
      .filter((n) => !allowed.includes(n))
    expect(
      unexpected,
      `new instructions since this test was written — check none mutate Config: ${unexpected.join(', ')}`,
    ).to.be.empty
  })

  /**
   * Ruling A (2026-08-12) removed every holding cost, and the removal has to be
   * provable rather than merely untested. The allowlist above only catches
   * instructions being *added*; this catches them coming back.
   *
   * It matters beyond tidiness: both instructions existed to move ENC out of a
   * holder's wallet without their signature, which is the one power the design
   * now claims it does not have. The claim is checkable, so check it.
   */
  it('has no rent or foreclosure instruction, and no rent parameters', () => {
    const names = h.program.idl.instructions.map((i) => i.name)
    for (const gone of ['settle_rent', 'settleRent', 'foreclose']) {
      expect(names, `${gone} is back in the IDL`).to.not.include(gone)
    }
    // The parameters go too — a Config field with no instruction behind it is
    // a rule the coin cannot enforce and a sentence the README would owe.
    const fields = JSON.stringify(h.program.idl.accounts ?? []) + JSON.stringify(h.program.idl.types ?? [])
    for (const gone of ['rentRatePerDayBps', 'rent_rate_per_day_bps', 'graceSeconds', 'grace_seconds', 'forecloseBounty', 'foreclose_bounty', 'rentAccrued', 'rent_accrued']) {
      expect(fields, `${gone} is still in the published interface`).to.not.include(gone)
    }
  })

  /**
   * The mock instructions are a compile-time choice, not a gated one.
   *
   * `set_mock_m2` could set the money supply by hand and `mock_fund` could move
   * vault ENC to a chosen wallet — either would be fatal to the coin's only
   * real claim, which is that its supply is not ours to choose.
   *
   * **Reads the committed file on purpose.** `h.program.idl` is whatever
   * `anchor build` last wrote into `target/`, which under `./stack test` is a
   * *mock* build — so asserting against it would test the wrong artifact and
   * fail for the right-looking wrong reason. `chain/idl/` is what ships, and
   * only a default build publishes there.
   */
  it('ships an interface with no mock instruction in it', () => {
    const shipped = JSON.parse(
      readFileSync(new URL('../idl/emperors_new_coin.json', import.meta.url), 'utf8'),
    ) as { instructions: { name: string }[] }
    const names = shipped.instructions.map((i) => i.name)
    expect(names, 'the shipped IDL is empty — run `./stack build`').to.not.be.empty
    for (const gone of ['set_mock_m2', 'setMockM2', 'mock_fund', 'mockFund']) {
      expect(names, `${gone} is in the SHIPPED IDL — a default build must not carry it`).to.not.include(
        gone,
      )
    }
  })

  /** Only the upgrade authority may bootstrap, so nobody can front-run us. */
  it('refuses a stranger', function () {
    if (strangerFailure === null) this.skip()
    expect(strangerFailure, 'a stranger was allowed to bootstrap the coin').to.match(
      /NotUpgradeAuthority|upgrade authority/i,
    )
  })

  // ── The money ─────────────────────────────────────────────────────────────

  it('creates ENC with six decimals and NO freeze authority', async () => {
    const mint = await getMint(h.connection, h.mintPda, undefined, TOKEN_PROGRAM_ID)
    expect(mint.decimals).to.equal(6)
    // Not optional: a freeze authority is a live BadCode key over other
    // people's coins, and it can only be removed by never having set it.
    expect(mint.freezeAuthority, 'ENC has a freeze authority').to.be.null
    // The mint authority stays forever, and stays a PDA. Scanners will flag ENC
    // as arbitrarily inflatable — correctly, and that is the joke.
    expect(mint.mintAuthority?.toBase58()).to.equal(h.vaultPda.toBase58())
  })

  it('puts 100% of the genesis supply in the vault', async function () {
    // Only true before anything moves tokens: sync_m2 retargets the supply, and
    // the faucet will start paying it away at T13.
    const printer = await h.program.account.printer.fetch(h.printerPda)
    if (printer.m2ReleaseDate.toNumber() !== 0) this.skip()
    expect((await h.supply()).toString()).to.equal(EXPECTED_SUPPLY.toString())
    // BadCode takes no allocation. At genesis there is nobody else.
    expect((await h.vaultBalance()).toString()).to.equal((await h.supply()).toString())
  })

  it('records a genesis release date the first sync can beat', async function () {
    const printer = await h.program.account.printer.fetch(h.printerPda)
    if (printer.m2ReleaseDate.toNumber() !== 0) this.skip()
    expect(printer.m2Value.toString()).to.equal(GENESIS_M2.toString())
    expect(printer.targetSupply.toString()).to.equal(EXPECTED_SUPPLY.toString())
  })

  // ── The assets ────────────────────────────────────────────────────────────

  it('creates exactly ten, all held by the Emperor', async function () {
    const config = await h.program.account.config.fetch(h.configPda)
    expect(config.initializedAssets).to.equal(ASSET_COUNT)

    const assets = await Promise.all(
      Array.from({ length: ASSET_COUNT }, (_, i) => h.program.account.asset.fetch(h.assetPda(i))),
    )
    assets.forEach((a, i) => expect(a.index).to.equal(i))

    // "All held by the Emperor" is a statement about *genesis*, and the whole
    // point of the auction is to stop it being true. On a ledger the auction
    // suite has already played, skip rather than assert something the design
    // deliberately falsifies — `./stack reset` if you want this to run.
    if (assets.some((a) => a.holder.toBase58() !== h.vaultPda.toBase58())) this.skip()
    for (const [i, a] of assets.entries()) {
      expect(a.holder.toBase58(), `asset ${i} is not in the vault`).to.equal(h.vaultPda.toBase58())
    }
  })

  it('mints each as a true NFT: supply 1, no mint authority', async () => {
    for (let i = 0; i < ASSET_COUNT; i++) {
      const mint = await getMint(h.connection, h.assetMintPda(i), undefined, TOKEN_2022_PROGRAM_ID)
      expect(mint.decimals, `asset ${i}`).to.equal(0)
      expect(mint.supply.toString(), `asset ${i}`).to.equal('1')
      // Dropped after minting the single unit, so no more can ever exist.
      expect(mint.mintAuthority, `asset ${i} can still be minted`).to.be.null
      const held = await h.connection.getTokenAccountBalance(
        h.vaultAta(h.assetMintPda(i), TOKEN_2022_PROGRAM_ID),
      )
      expect(held.value.amount, `asset ${i}`).to.equal('1')
    }
  })

  it('carries metadata a wallet can render', async () => {
    for (let i = 0; i < ASSET_COUNT; i++) {
      const meta = await getTokenMetadata(
        h.connection,
        h.assetMintPda(i),
        undefined,
        TOKEN_2022_PROGRAM_ID,
      )
      expect(meta, `asset ${i} has no metadata`).to.not.be.null
      expect(meta!.name).to.equal(assetSpec(i).name)
      expect(meta!.uri).to.equal(assetSpec(i).uri)
    }
  })

  /**
   * The opposite of what T8 asserted, and deliberately so.
   *
   * The permanent delegate existed to make a forced sale possible. Ruling A
   * (2026-08-12) removed the forced sale, and the ten NFTs now never leave the
   * program's custody — so the power would have been dead weight while costing
   * the heaviest flag a risk scanner issues. Its absence is what makes "no
   * token leaves any wallet without its owner's signature" structural.
   */
  it('gives nobody permanent delegate over any asset', async () => {
    for (let i = 0; i < ASSET_COUNT; i++) {
      const mint = await getMint(h.connection, h.assetMintPda(i), undefined, TOKEN_2022_PROGRAM_ID)
      expect(getPermanentDelegate(mint), `asset ${i} has a permanent delegate`).to.be.null
    }
  })

  it('refuses an eleventh asset, and a repeat of an existing one', async () => {
    for (const index of [10, 255, 0]) {
      const why = await h.failureOf(() =>
        h.program.methods
          .initAsset(index, 'x', 'x', 'x', new BN(1))
          .accounts(initAssetAccounts(h, index))
          .rpc(),
      )
      expect(why, `init_asset(${index}) was allowed`).to.not.equal('')
    }
  })

  /**
   * The reason `initialize` and `init_asset` are separate: all ten assets in one
   * call is roughly forty accounts, past both the compute and transaction-size
   * limits.
   */
  it('fits one asset in one transaction, well under the compute limit', async function () {
    // Only measurable when this run created it. solana-test-validator does not
    // index getSignaturesForAddress for these accounts, so on a warm ledger
    // there is nothing to look up — skip rather than pretend.
    const sig = created.get(ASSET_COUNT - 1)
    if (sig === undefined) this.skip()
    const tx = await h.history.getTransaction(sig, { maxSupportedTransactionVersion: 0 })
    const consumed = (tx?.meta?.logMessages ?? [])
      .map((l) => /consumed (\d+) of (\d+) compute units/.exec(l))
      .find(Boolean)
    expect(consumed, 'no compute-unit log found').to.not.be.undefined
    const used = Number(consumed![1])
    expect(used, `init_asset used ${used} CU`).to.be.lessThan(200_000)
  })
})

File: chain/programs/emperors-new-coin/src/instructions/initialize.rs
//! Bring the coin into existence. Once, and never again.
//!
//! Creates the ENC mint, the vault that holds every token at genesis, the
//! immutable `Config`, and the `Printer` that records what the Fed last said.
//!
//! **Who may call this.** Only the program's current upgrade authority, checked
//! against the loader's own `ProgramData` account. That deliberately introduces
//! no new key: the upgrade authority already exists, it is burned at T22, and
//! gating on it closes the one real front-running hole — otherwise a stranger
//! could call `initialize` between deploy and our own first transaction and set
//! every economic parameter themselves.
use anchor_lang::prelude::*;
use anchor_spl::associated_token::AssociatedToken;
use anchor_spl::token::{mint_to, Mint, MintTo, Token, TokenAccount};

use crate::errors::EncError;
use crate::math::{target_supply, BPS};
use crate::state::*;

/// The economic parameters, chosen once and then unchangeable.
///
/// Supplied by the caller rather than hardcoded so `chain/params.genesis.json`
/// stays the single source of truth and T15 can replace the placeholders
/// without touching Rust. Everything here is validated below; nothing here has
/// a setter anywhere in the program.
#[derive(AnchorSerialize, AnchorDeserialize, Clone, Debug)]
pub struct InitializeParams {
    /// The Switchboard feed this program will ever accept a quote from.
    pub expected_feed_id: [u8; 32],
    /// Base units of ENC per unit of `m2_value`.
    pub k: u64,
    pub faucet_alpha_bps: u16,
    pub floor_bps: u16,
    pub welcome_grant: u64,
    pub grants_per_epoch: u16,
    pub term_seconds: i64,
    pub epoch_seconds: i64,
    pub retirement_silence_seconds: i64,
    pub max_change_bps: u16,

    /// Who holds the editor's pen at genesis, or `None` for a paper that is
    /// feral from birth.
    ///
    /// **Not an economic parameter**, which is why it is not in
    /// `params.genesis.json` alongside the rest — it is a key, chosen at
    /// deployment, and the only one this program ever accepts. It can be
    /// rotated afterwards by its holder and broken by its holder, and by
    /// nobody else, ever.
    pub editor: Option<Pubkey>,
}

pub fn handler(ctx: Context<Initialize>, params: InitializeParams) -> Result<()> {
    validate(&params)?;

    let clock = Clock::get()?;
    let genesis_supply = target_supply(GENESIS_M2_VALUE, params.k)?;

    // ── The rules ───────────────────────────────────────────────────────────
    let config = &mut ctx.accounts.config;
    config.mint = ctx.accounts.mint.key();
    config.vault = ctx.accounts.vault.key();
    config.expected_feed_id = params.expected_feed_id;
    config.k = params.k;
    config.enc_decimals = ENC_DECIMALS;
    config.faucet_alpha_bps = params.faucet_alpha_bps;
    config.floor_bps = params.floor_bps;
    config.welcome_grant = params.welcome_grant;
    config.grants_per_epoch = params.grants_per_epoch;
    config.term_seconds = params.term_seconds;
    config.epoch_seconds = params.epoch_seconds;
    config.retirement_silence_seconds = params.retirement_silence_seconds;
    config.max_change_bps = params.max_change_bps;
    config.initialized_assets = 0;
    config.retired = false;
    config.editor = params.editor;
    config.bump = ctx.bumps.config;

    // ── What the Fed last said ──────────────────────────────────────────────
    let printer = &mut ctx.accounts.printer;
    printer.m2_value = GENESIS_M2_VALUE;
    // Zero, not "now": the anti-double-mint guard requires the release date to
    // strictly advance, and any real Fed release is later than the epoch. A
    // genesis date of `now` would reject every release published before today.
    printer.m2_release_date = 0;
    printer.last_sync_slot = clock.slot;
    // The retirement clock starts **now**, not at zero. Left at zero the coin
    // would be decades overdue for retirement the instant it was born, and the
    // first stranger to look could end it before the first sync.
    printer.last_sync_at = clock.unix_timestamp;
    printer.target_supply = genesis_supply;
    printer.bump = ctx.bumps.printer;

    // ── The money ───────────────────────────────────────────────────────────
    // Every token goes to the vault. There is nobody else yet, and BadCode
    // takes no allocation — the only way out of the vault is the faucet, which
    // makes it the only route into the economy at all.
    let vault_seeds: &[&[u8]] = &[VAULT_SEED, &[ctx.bumps.vault]];
    mint_to(
        CpiContext::new_with_signer(
            ctx.accounts.token_program.key(),
            MintTo {
                mint: ctx.accounts.mint.to_account_info(),
                to: ctx.accounts.vault_token_account.to_account_info(),
                authority: ctx.accounts.vault.to_account_info(),
            },
            &[vault_seeds],
        ),
        genesis_supply,
    )?;

    msg!(
        "genesis: M2 {} -> {} base units, all of it in the vault",
        GENESIS_M2_VALUE,
        genesis_supply
    );
    Ok(())
}

/// Refuse parameters that are nonsense before they become permanent.
///
/// There is no instruction to fix any of this later, so anything caught here is
/// caught forever and anything missed here is permanent.
fn validate(p: &InitializeParams) -> Result<()> {
    require!(p.k > 0, EncError::InvalidRate);
    // A share of something cannot exceed the whole of it.
    require!(p.floor_bps as u128 <= BPS, EncError::InvalidRate);
    require!(p.faucet_alpha_bps as u128 <= BPS, EncError::InvalidRate);
    // A term of zero (or less) would expire the instant it began, so every
    // asset would sit permanently settleable and no bid could ever be placed.
    require!(p.term_seconds > 0, EncError::InvalidRate);
    // An epoch of zero would divide the clock by nothing and pin the whole
    // faucet to epoch 0 forever — one pot, once, and never another.
    require!(p.epoch_seconds > 0, EncError::InvalidRate);
    // A silence window of zero (or less) would let the first passer-by retire
    // the coin in the same block it was born in. The flag is irreversible on a
    // non-upgradeable program, so this is the one validation here whose absence
    // would be immediately fatal rather than eventually embarrassing.
    require!(p.retirement_silence_seconds > 0, EncError::InvalidRate);
    // A cap of zero would make every walk a one-unit step, so a real move would
    // take billions of transactions to land — the peg alive in principle and
    // frozen in practice.
    require!(p.max_change_bps > 0, EncError::InvalidRate);
    // An editor of the all-zero key would leave `editor.is_some()` true while
    // nobody could ever sign for it — a pen the account claims exists and no
    // hand holds. `None` says that honestly; this refuses to say it by
    // accident, exactly as `pass_the_pen` does.
    require!(
        p.editor != Some(Pubkey::default()),
        EncError::NotTheEditor
    );
    // The genesis supply must fit, or the program is born broken.
    target_supply(GENESIS_M2_VALUE, p.k)?;
    Ok(())
}

#[derive(Accounts)]
pub struct Initialize<'info> {
    /// The program's upgrade authority, and nobody else.
    #[account(mut)]
    pub authority: Signer<'info>,

    /// CHECK: verified against `program_data` below, which is the loader's own
    /// record of who may upgrade this program.
    #[account(constraint = program.programdata_address()? == Some(program_data.key()))]
    pub program: Program<'info, crate::program::EmperorsNewCoin>,

    #[account(constraint = program_data.upgrade_authority_address == Some(authority.key()) @ EncError::NotUpgradeAuthority)]
    pub program_data: Account<'info, ProgramData>,

    /// `init` is what makes this callable exactly once: the second attempt
    /// fails because the account already exists.
    #[account(
        init,
        payer = authority,
        space = 8 + Config::INIT_SPACE,
        seeds = [CONFIG_SEED],
        bump,
    )]
    pub config: Account<'info, Config>,

    #[account(
        init,
        payer = authority,
        space = 8 + Printer::INIT_SPACE,
        seeds = [PRINTER_SEED],
        bump,
    )]
    pub printer: Account<'info, Printer>,

    /// CHECK: a signer-only PDA. It holds no data of its own; it is the mint
    /// authority, the landlord and the faucet, and it signs by seed.
    #[account(seeds = [VAULT_SEED], bump)]
    pub vault: UncheckedAccount<'info>,

    /// ENC itself. Classic SPL so every DEX can list it.
    ///
    /// **Freeze authority is `None`, and that is not optional.** A freeze
    /// authority is a live key over other people's coins; omitting the
    /// constraint is what sets it to `None` at creation, irreversibly.
    /// The mint authority stays on the vault PDA forever, so scanners will
    /// report ENC as arbitrarily inflatable — which is true, and is the joke.
    #[account(
        init,
        payer = authority,
        mint::decimals = ENC_DECIMALS,
        mint::authority = vault,
        seeds = [MINT_SEED],
        bump,
    )]
    pub mint: Account<'info, Mint>,

    #[account(
        init,
        payer = authority,
        associated_token::mint = mint,
        associated_token::authority = vault,
    )]
    pub vault_token_account: Account<'info, TokenAccount>,

    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub system_program: Program<'info, System>,
    pub rent: Sysvar<'info, Rent>,
}
